Java doc: "Ray Core Walkthrough" page (#8595)
ray-project/ray: doc/source/walkthrough.rst

This walkthrough will overview the core concepts of Ray:

> .. tabs::

Reviewer: Overall this is great work - good to see the tabs and think they will add a lot to the community.

With respect to some guiding principles having worked on Spark is that targeting having the fewest number of tab groups. It becomes very complicated to maintain over time and to enable other contributors. 

The one major request I would have is that It'd be good to minimize the contents of each tab group to only include the minimum amount of information as possible - even if that involves a little reorganization of the information that we currently have.

Remote functions (Tasks)
------------------------

Ray enables arbitrary functions to be executed asynchronously. These asynchronous Ray functions are called "remote functions". Here is an example.

.. tabs::
  .. group-tab:: Python

    .. code:: python

>       # A regular Python function.

Author: I reorganized this part. I used to compare regular python function and remote function in different aspects. Now it just directly explains the features of remote functions. 

This is to make Python and Java examples more aligned. Because in Java, there is no such an entity of "remote function". You just take a regular Java function, and invoke it by `Ray.task(func).remote()`

I think the comparison was unnecessary. Because all users should already know regular Python functions are called by `regular_function()`.

      // a task that will be executed on a worker process.
      ObjectRef<Integer> res = Ray.task(MyRayApp::myFunction).remote();

      // The result can be retrieved with ``ray.get``.
      Assert.assertTrue(res.get() == 1);

      public static int slowFunction() throws InterruptedException {
        TimeUnit.SECONDS.sleep(10);
        return 1;
      }

      // Invocation of Ray remote functions happen in parallel.
      // All computation is performed in the background, driven by Ray's internal event loop.
      for(int i = 0; i < 4; i++) {
        // This doesn't block.
        Ray.task(MyRayApp::slowFunction).remote();
      }

.. _ray-object-refs:

Passing object refs to remote functions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Object refs** can also be passed into remote functions. When the function actually gets executed, **the argument will be a retrieved as a regular object**. For example, take this function:

.. tabs::
  .. code-tab:: python

    @ray.remote
    def function_with_an_argument(value):
        return value + 1

    obj_ref1 = my_function.remote()
    assert ray.get(obj_ref1) == 1

>     # You can pass an `ObjectRef` as an argument to another Ray remote function.

Reviewer 2: ```suggestion
    # You can pass an object ref as an argument to another Ray remote function.
```

Since `ObjectRef` is not in public API.

    obj_ref2 = function_with_an_argument.remote(obj_ref1)
    assert ray.get(obj_ref2) == 2

  .. code-tab:: java

    public static int functionWithAnArgument(int value) {
      return value + 1;
    }

    ObjectRef<Integer> objRef1 = Ray.task(MyRayApp::myFunction).remote();
    Assert.assertTrue(objRef1.get() == 1);

>     // You can pass an `ObjectRef` as an argument to another Ray remote function.

Reviewer 2: ```suggestion
    // You can pass an object ref as an argument to another Ray remote function.
```

To keep it consistent.

Remote functions (Tasks)
------------------------

.. tabs::

>   .. group-tab:: Python

Reviewer 3: This tab is pretty huge and contains a lot of duplicated text. Can we break it down into smaller tabs and have more common text shared between Python and Java?

Reviewer: +1, Can you please make sure that every tab group is as small as it can possibly be?

Author: Updated

        * If specifying GPUs, Ray does provide isolation in forms of visible devices
          (setting the environment variable ``CUDA_VISIBLE_DEVICES``), but it is the
          task's responsibility to actually use the GPUs (e.g., through a deep
          learning framework like TensorFlow or PyTorch).

    .. code-block:: java

      Map<String, Double> resources = ImmutableMap.of("CPU", 4, "GPU", 2);
      CallOptions options = new CallOptions.Builder()
        .setResources(resources).createCallOptions();
      Ray.call(MyRayApp::foo, callOptions);

    The resource requirements of a task have implications for the Ray's scheduling
    concurrency. In particular, the sum of the resource requirements of all of the
    concurrently executing tasks on a given node cannot exceed the node's total
    resources.

    Below are more examples of resource specifications:

    .. code-block:: java

      // Ray aslo supports fractional and custom resources.
      Map<String, Double> resources = ImmutableMap.of("GPU", 0.5, "Custom", 1);
      CallOptions options = new CallOptions.Builder()
        .setResources(resources).createCallOptions();
      Ray.call(MyRayApp::foo, callOptions);

>     .. TODO(hchen): Implment `Ray.cancel` in Java and add doc here.

Reviewer 2: ```suggestion
    .. note:: Task canceling isn't supported in Java yet.
```

    An **object ID** is essentially a unique ID that can be used to refer to a
    remote object. If you're familiar with futures, our object IDs are conceptually
    similar.

    Object IDs can be created in multiple ways.

      1. They are returned by remote function calls.
      2. They are returned by ``ray.put``.

    .. code-block:: python

        y = 1
        object_id = ray.put(y)

    .. autofunction:: ray.put
        :noindex:

  .. group-tab:: Java

    In Ray, we can create and compute on objects. We refer to these objects as **remote objects**, and we use ``RayObject``-s to refer to them. Remote objects are stored in `shared-memory <https://en.wikipedia.org/wiki/Shared_memory>`__ **object stores**, and there is one object store per node in the cluster. In the cluster setting, we may not actually know which machine each object lives on.

>     An ``RayObject`` is essentially a reference to a remote object. If you're familiar with futures, our ``RayObject``-s are conceptually similar.

Reviewer 2: ```suggestion
    A ``RayObject`` is essentially a reference to a remote object. If you're familiar with futures, our ``RayObject``-s are conceptually similar.
```

      CallOptions options = new CallOptions.Builder()
        .setResources(resources).createCallOptions();
      Ray.call(MyRayApp::foo, callOptions);

Multiple returns
~~~~~~~~~~~~~~~~

.. tabs::
  .. group-tab:: Python

    Python remote functions can return multiple object IDs.

    .. code-block:: python

      @ray.remote(num_return_vals=3)
      def return_multiple():
          return 1, 2, 3

      a_id, b_id, c_id = return_multiple.remote()

  .. group-tab:: Java

    Java remote functions doesn't support returning multiple objects.

Cancelling tasks
~~~~~~~~~~~~~~~~

.. tabs::
  .. group-tab:: Python

    Remote functions can be canceled by calling ``ray.cancel`` on the returned Object ID. Remote actor functions can be stopped by killing the actor using the ``ray.kill`` interface.

    .. code-block:: python

      @ray.remote
      def blocking_operation():
          time.sleep(10e6)
          return 100

      obj_id = blocking_operation.remote()
      ray.cancel(obj_id)

    .. autofunction:: ray.cancel
        :noindex:

  .. group-tab:: Java

    Cancelling tasks hasn't been implemented in Java yet.

Objects in Ray
--------------

.. tabs::

>   .. group-tab:: Python

Reviewer: Since the semantics are the same here, it's just minor changes to the API calls, can we just add a code block for the call as opposed to changing the whole paragraph?

Having to maintain huge groups of text between the two is going to be very difficult.

For instance, here it's probably easier to say 

>We refer to these objects as **remote objects**, and we use **object IDs** to refer to them. Remote objects are stored ...

as 
> We refer to these objects as **remote objects**, and we use **object IDs** (or ``RayObject`` in Java) to refer to them. Remote objects are stored....

If the current node's object store does not contain the object, the object is downloaded.

.. tabs::
  .. group-tab:: Python

    If the object is a `numpy array <https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html>`__
    or a collection of numpy arrays, the ``get`` call is zero-copy and returns arrays backed by shared object store memory.
    Otherwise, we deserialize the object data into a Python object.

    .. code-block:: python

      # Get the value of one object ref.
      obj_ref = ray.put(1)
      assert ray.get(obj_ref) == 1

      # Get the values of multiple object refs in parallel.
      assert ray.get([ray.put(i) for i in range(3)]) == [0, 1, 2]

      # You can also set a timeout to return early from a ``get`` that's blocking for too long.
      from ray.exceptions import RayTimeoutError

      @ray.remote
      def long_running_function()
          time.sleep(8)

      obj_ref = long_running_function.remote()
      try:
          ray.get(obj_ref, timeout=4)
      except RayTimeoutError:
          print("`get` timed out.")

  .. group-tab:: Java

    .. code-block:: java

>       // Get the value of one object ref.
>       ObjectRef<Integer> objRef = Ray.put(1);
>       // Get the value of one object ref.
>       ObjectRef<Integer> objRef = Ray.put(1);

Reviewer 2: ```suggestion
      // Get the value of one object ref.
      ObjectRef<Integer> objRef = Ray.put(1);
```

    .. code:: java

      List<RayObject<Integer>> rayObjectList = new ArrayList<>();
      for (int i = 0; i < 4; i++) {
        rayObjectList.add(Ray.call(MyRayApp::foo));
      }
      // Fetch multiple objects in parallel.
      List<Integer> results = Ray.get(rayObjectList);
      Assert.assertEquals(results, ImmutableList.of(1, 1, 1, 1));

    If an exception occurs while executing a remote function, the exception
    will be stored in the ``RayObject``. When fetching the result, a `RayTaskException`
    will be thrown.

    .. code:: java

      public static int badMethod() {
        throw new RuntimeException("An expected exception occurred.");
      }

      try {
        rayObject.get();
      } catch (RayTaskException e) {
        Throwable rootCause = e.getCause();
        Assert.assertTrue(rootCause instanceof RuntimeException);
      }

    After launching a number of tasks, you may want to know which ones have
    finished executing. This can be done with ``Ray.wait``. The function
    works as follows.

    .. code:: java

        WaitResult<Integer> waitResult = Ray.wait(rayObjectList, /*num_returns=*/1, /*timeoutMs=*/1000);

>         System.out.println(waitResult.getReady());  // List of ready objects.

Reviewer 2: Will this print a human-readable output?

    being evicted, you can call ``ray.get`` and store their values instead. Numpy array
    objects cannot be evicted while they are mapped in any Python process. You can also
    configure `memory limits <memory-management.html>`__ to control object store usage by
    actors.

    .. note::

        Objects created with ``ray.put`` are pinned in memory while a Python reference
        to the object ID returned by the put exists. This only applies to the specific
        ID returned by put, not IDs in general or copies of that IDs.

  .. group-tab:: Java

    When the object store gets full, objects will be evicted to make room for new objects.
    This happens in approximate LRU (least recently used) order. To avoid objects from
    being evicted, you can call ``RayObject::get`` or ``Ray.get`` and store their values instead.
    You can also configure `memory limits <memory-management.html>`__ to control object store usage by
    actors.

    .. note::

        Objects created with ``Ray.put`` are pinned in memory while a Java reference
        to the ``RayObject`` returned by the put exists. This only applies to the specific

>         ``RayObject``-s returned by put, not ``RayObject``-s in general or copies of thoses ``RayObject``-s.

Reviewer 2: Pinning in Java is not implemented yet. We need to rewrite this section for Java.

Reviewer: Since it seems like there are going to be some limitations/significant differences for Ray on Java across the board, I recommend that we put all the java information in its own page. In doing so we make it extremely simple to reference for users using Java / wanting to get started and we get a single place that we can call everything out in terms of advantages or limitations.

This also makes sense since Tune / RLlib / SGD / etc. aren't going to be supported.

Also, I'm going to do a large scale overall of the core docs to make it easier to search and organize. If you like, I am more than happy to include the Java docs information in my PR.

Reviewer: I thought about this a bit more and think this is all a great addition! Please ignore my previous comment and let's get this merged for the next release!

    operator. We can then call ``ray.get`` on the object ID to retrieve the actual
    value.

    .. code-block:: python

      obj_id = a1.increment.remote()
      ray.get(obj_id) == 1

  .. group-tab:: Java

    Actors extend the Ray API from functions (tasks) to classes. The ``Ray.createActor``
    is used to create actors from normal Java classes. An actor is essentially a stateful worker.
    Unlike Python, multiple Java actors may share one JVM process, in order to reduce JVM's memory
    overhead. But this is transparent to normal users.

    .. code-block:: java

      // A normal Java class.
      public class Counter {

        private int value = 0;

        public int increment() {
          this.value += 1;
          return this.value;
        }
      }

      // Create an actor from this class.
      // `Ray.createActor` takes a factory method that can produce
      // a `Counter` object. Here, we pass `Counter`'s constructor
      // as the argument.
      RayActor<Counter> actor = Ray.createActor(Counter::new);

    You can specify resource requirements in Actors too (see the `Actors section

>     <actors.html>`__ for more details).

Reviewer 2: Reuse this passage.

      Ray.createActor(Counter::new, options);

    We can interact with the actor by calling its methods with the ``RayActor::call``
    method, which also returns a ``RayObject``.

    .. code-block:: java

      RayObject<Integer> result = actor.call(Counter::increment);
      Assert.assertEqual(result.get(), 1);

Methods called on different actors can execute in parallel, and methods called on the same actor are executed serially in the order that they are called. Methods on the same actor will share state with one another, as shown below.

.. tabs::

  .. group-tab:: Python

    .. code-block:: python

      # Create ten Counter actors.
      counters = [Counter.remote() for _ in range(10)]

      # Increment each Counter once and get the results. These tasks all happen in
      # parallel.
      results = ray.get([c.increment.remote() for c in counters])
      print(results)  # prints [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

      # Increment the first Counter five times. These tasks are executed serially
      # and share state.
      results = ray.get([counters[0].increment.remote() for _ in range(5)])
      print(results)  # prints [2, 3, 4, 5, 6]

  .. group-tab:: Java

>     .. code-block:: java

Reviewer 2: FYI, you can use `.. code-tab:: java` instead of `.. group-tab:: Java` and `.. code-block:: java`. See https://github.com/djungelorm/sphinx-tabs#code-tabs.

    will be stored in the ``RayObject``. When fetching the result, a `RayTaskException`
    will be thrown.

    .. code:: java

      public static int badMethod() {
        throw new RuntimeException("An expected exception occurred.");
      }

      try {
        rayObject.get();
      } catch (RayTaskException e) {
        Throwable rootCause = e.getCause();
        Assert.assertTrue(rootCause instanceof RuntimeException);
      }

After launching a number of tasks, you may want to know which ones have
finished executing. This can be done with ``wait``. The function
works as follows.

.. tabs::
  .. group-tab:: Python

    .. code:: python

        ready_ids, remaining_ids = ray.wait(object_ids, num_returns=1, timeout=None)

    .. autofunction:: ray.wait
        :noindex:

  .. group-tab:: Java

    .. code:: java

        WaitResult<Integer> waitResult = Ray.wait(rayObjectList, /*num_returns=*/1, /*timeoutMs=*/1000);
        System.out.println(waitResult.getReady());  // List of ready objects.
        System.out.println(waitResult.getUnready());  // list of unready objects.

Object Eviction
---------------

.. tabs::

>   .. group-tab:: Python

Reviewer: Since the semantics are the same here, it's just minor changes to the API calls, can we just add a code block for the call as opposed to changing the whole paragraph?

Having to maintain huge groups of text between the two is going to be very difficult.
